Add relro-level tests

kyrias opened the pull request:
The `relro-level` debugging flag was added in #43170 which was merged in July 2017.  This PR moves this flag to be a proper codegen flag.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

nikomatsakis commented:
This is basically a stabilization. Is there a tracking issue for this feature?

nikomatsakis commented:
I know nothing about this, so it's hard for me to judge if we are happy enough with its implementation to be ready to stabilize. =) 

cc @rust-lang/compiler

kyrias commented:
No tracking issue.  This is essentially just a flag to be able to *disable* RELRO, or use only partial RELRO.  Full RELRO has been on by default as of #43170 being merged, but @alexcrichton wanted the flag to disable it unstable for a while at the time.

nikomatsakis commented:
Ok. Usually for a stabilization I like to see a little report that shows the test cases that show that this feature is indeed working as expected -- based on the the diffs in this PR, I take it that there are no tests at all for this (i.e., nothing had to be changed from `-Z` to `-C`). Do you think you'd be up for adding one?

kyrias commented:
Not really sure how to test this from rust honestly.  To actually check that it is functioning properly we'd essentially need to actually do a compile and then parse the ELF binary and check for the `GNU_RELRO` header and `BIND_NOW` sections.  Not sure if there's any nice way to do that though?

nikomatsakis commented:
@kyrias usually we would make a `run-make` test for that sort of thing

nikomatsakis commented:
Maybe it's not worth the trouble, I don't know, but I am a big believer in "if it ain't tested, it's broken".

kyrias commented:
Interesting, didn't know about `run-make` tests, I'll give it a shot.  (Is there any good docs for getting more details about the different parts of the compiler tests, and how to write the different types of tests.)

Is there any specific list of dependencies that are guaranteed to exist when they're run?

kyrias commented:
So I figured that when you specify `relro-level=off` you probably want to tell the linker to explicitly not use RELRO rather than just have the compiler not pass on the flag at all, so I added that as well, and then a test for this case.

Since this is a very small change I just added it in this PR, but if you would prefer I can split it out and file a PR for it separately.

bors commented:
:umbrella: The latest upstream changes (presumably #48531) made this pull request unmergeable. Please resolve the merge conflicts.

pietroalbini commented:
Hi @kyrias, ping from the release team! @alexcrichton asked you more details about the stabilization in https://github.com/rust-lang/rust/pull/48388#issuecomment-368573113, can you please reply so we can move forward with this PR?

kyrias commented:
Hey, sorry for the delay, been away for a week.  Now that I think about it I don't really have any specific plans for it, I just thought that it can be useful to be able to disable it in some rare circumstances.  Since the executables end up being slightly bigger, disabling RELRO can be useful for extremely low memory targets.  But other than that I guess there aren't anything in specific I can think off.

alexcrichton commented:
Ah ok thanks for the info @kyrias! In that case I wonder if it's best left unstable for now? That way it's there to tinker with if necessary but otherwise we may want to hold out as it may inform how we stabilize this.

kyrias commented:
Yeah, might make sense until the problems become more real.  I think I'm mostly just bothered by the fact that `-Z` is documented as debug options everywhere, which bothers me since it's not actually a debug option, heh.

Anyway, I could drop the move of the option and just keep the tests and the `relro-level=off` fix.

alexcrichton commented:
Sounds great to me!

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 54b68b6 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 54b68b656913baa760dc9daf75796e84172f810a -->

bors commented:
:hourglass: Testing commit 54b68b656913baa760dc9daf75796e84172f810a with merge 08fb75096f0aacfd05076f8bb674542aee0e9d01...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6601)

kennytm commented:
The `-z norelro` flag is unrecognized on i686-pc-windows-gnu. This option is ELF-only, and should not be passed when the target is PE (Windows) or Mach-O (macOS).

```
error: linking with `gcc` failed: exit code: 1
  |
  = note: "gcc" ...... "-Wl,-z,norelro" ......
  = note: C:/projects/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/6.3.0/../../../../i686-w64-mingw32/bin/ld.exe: unrecognized option '-z'
          C:/projects/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/6.3.0/../../../../i686-w64-mingw32/bin/ld.exe: use the --help option for usage information
          collect2.exe: error: ld returned 1 exit status
```

nikomatsakis commented:
@kyrias `-Z` should really be marked as "unstable options"

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 1dbce4b has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 1dbce4b0afb8f76eb7cd7aad0e3c34d4f7fc2274 -->

bors commented:
:hourglass: Testing commit 1dbce4b0afb8f76eb7cd7aad0e3c34d4f7fc2274 with merge 7f6687d357f136197ea8285137bfddb7cd99c231...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/351415132?utm_source=github_status&utm_medium=notification)

pietroalbini commented:
@bors retry #48866

bors commented:
:hourglass: Testing commit 1dbce4b0afb8f76eb7cd7aad0e3c34d4f7fc2274 with merge 211c155e8eb32ad9e147edf35471e6a2c73c30b5...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/351585167?utm_source=github_status&utm_medium=notification)

kennytm commented:
@bors retry 

Some OpenSSL installation error, should be already fixed by #48901.

bors commented:
:hourglass: Testing commit 1dbce4b0afb8f76eb7cd7aad0e3c34d4f7fc2274 with merge 2f0e6a3ba585e00205805d6cb9fdcbe4b1d1be63...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6628), [status-travis](https://travis-ci.org/rust-lang/rust/builds/351694190?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 2f0e6a3ba585e00205805d6cb9fdcbe4b1d1be63 to master...
